packages/helpermodules/simcount.py

```diff
@@ -0,0 +1,180 @@
+    """ emulate import export
+
+    Parameters
+    ----------
+    present_power_all: float
+        aktuelle Leistung
+    topic: str "openWB/set/counter/0/"
+        Topic, an das gepublished werden soll
+    data:  data.data.counter_data[item].data["set"]
+        Daten aus dem data-Modul auf die lesen zugegriffen wird.
+    Return
+    ------
+    imported: importierte Energie
+    exported: exportierte Energie
+    """
+    try:
+        sim_timestamp = time.time()
+        watt1 = 0
+        seconds1 = 0.0
+        start_new = True
+        if ramdisk == True:
+            if os.path.isfile('/var/www/html/openWB/ramdisk/'+pref+'sec0'): 
+                f = open('/var/www/html/openWB/ramdisk/'+pref+'sec0', 'r')
+                seconds1=float(f.read())
+                f.close()
+                f = open('/var/www/html/openWB/ramdisk/'+pref+'wh0', 'r')
+                watt1=int(float(f.read()))
+                f.close()
+                try:
+                    f = open('/var/www/html/openWB/ramdisk/'+pref+'watt0pos', 'r')
+                    wattposh=int(float(f.read()))
+                    importtemp = wattposh
+                    f.close()
+                except:
+                    importtemp = restore("watt0pos")
+
+                try:
+                    f = open('/var/www/html/openWB/ramdisk/'+pref+'watt0neg', 'r')
+                    wattnegh=int(f.read())
+                    exporttemp = wattnegh
+                    f.close()
+                except:
+                    exporttemp = restore("watt0neg")
+                f = open('/var/www/html/openWB/ramdisk/'+pref+'sec0', 'w')
+                value1 = "%22.6f" % sim_timestamp
+                f.write(str(value1))
+                f.close()
+                f = open('/var/www/html/openWB/ramdisk/'+pref+'wh0', 'w')
+                f.write(str(present_power_all))
+                f.close()
+                start_new = False
+        else:
+            if "sim_timestamp" in data:
+                seconds1 = float(data["sim_timestamp"])
+                watt1 = int(data["present_power_all"])
+                if "present_imported" in data:
+                    wattposh = int(data["present_imported"])
+                else:
+                    wattposh = 0
+                if "present_exported" in data:
+                    wattnegh = int(data["present_exported"])
+                else:
+                    wattnegh = 0
+                value1 = "%22.6f" % sim_timestamp
+                pub.pub(topic+"module/simulation/sim_timestamp", value1)
+                pub.pub(topic+"module/simulation/present_power_all", present_power_all)
+                start_new = False
+            
+        if start_new == False:
+            seconds1 = seconds1+1
+            deltasec = sim_timestamp - seconds1
+            deltasectrun = int(deltasec * 1000) / 1000
+            stepsize = int((present_power_all-watt1)/deltasec)
+            while seconds1 <= sim_timestamp:
```

Comment:
Ich finde diese Schleife (und eigentlich den ganzen Block von Zeile 92 bis 115 eher schwierig zu verstehen und auch etwas umständlich. Warum die Schleife? Ich habe darüber eine Weile nachgedacht und bin zu dem Ergebnis gekommen, dass hier versucht wird den Leistungsverlauf zu integrieren. Das geht aber auch einfacher und genauer. Hier ein Gegenvorschlag von mir:
```python
def calculate_import_export(seconds_since_previous: Number, power1: Number, power2: Number) -> Tuple[Number, Number]:
    power_low = min(power1, power2)
    power_high = max(power1, power2)
    def energy_function(seconds): return .5 * seconds ** 2 + power_low * seconds

    power_zero_seconds = -power_low * seconds_since_previous / (power_high - power_low)
    energy_total = energy_function(seconds_since_previous)
    if 0 < power_zero_seconds < seconds_since_previous:
        energy_exported = energy_function(power_zero_seconds)
        result = energy_total - energy_exported, energy_exported * -1
        return result
    return (energy_total, 0) if energy_total >= 0 else (0, -energy_total)
```
Ich habe auch ein paar Tests dazu geschrieben:
```python
Number = Union[int, float]

class TestCounterSimulation(unittest.TestCase):
    def test_energy_calculation(self: TestCase):
        cases = [
            # Ascending from zero:
            {"seconds": 1, "power_previous": 0, "power_present": 1, "expected_energy": (.5, 0)},
            # Ascending from more than zero:
            {"seconds": 1, "power_previous": 10, "power_present": 11, "expected_energy": (10.5, 0)},
            # Ascending to zero:
            {"seconds": 1, "power_previous": -1, "power_present": 0, "expected_energy": (0, .5)},
            # Ascending to less than zero:
            {"seconds": 1, "power_previous": -11, "power_present": -10, "expected_energy": (0, 10.5)},
            # Ascending from negative to positive
            {"seconds": 2, "power_previous": -1, "power_present": 1, "expected_energy": (.5, .5)},
            # Ascending from negative to positive with asymmetric intersection
            {"seconds": 10, "power_previous": -9, "power_present": 1, "expected_energy": (.5, 40.5)},
        ]
        for case in cases:
            with self.subTest(f"{case['power_previous']} -> {case['power_present']} ({case['seconds']}seconds)"):
                self.assertEqual(calculate_import_export(
                    case["seconds"], case["power_previous"], case["power_present"]), case["expected_energy"]
                )
            # Also repeat the tests by exchanging power_present and power_previous so that the values a descending
            # instead of ascending. This should not change the result:
            with self.subTest(f"{case['power_present']} -> {case['power_previous']} ({case['seconds']}seconds)"):
                self.assertEqual(calculate_import_export(
                    case["seconds"], case["power_present"], case["power_previous"]), case["expected_energy"]
                )


if __name__ == '__main__':
    unittest.main()
```